### DISCUSSION TITLE
Can retrofit support template well?

### DISCUSSION THREAD

**Discussion by markwang1971** [square/retrofit#3529] [General]
interface GetRadioListApi {
    @GET(SUB_URL)
    suspend fun  getRadioList(@QueryMap queryMap: BaseRequest): RadioListResponse
    data  class RadioListResponse(val msg: String, val ret: Int, val sub_ret: Int)
}
publicRadioListApi.getRadioList(params)  does work well. 

But the following does not work.
interface GetRadioListApi {
    @GET(SUB_URL)
    suspend fun  < T >  getRadioList(@QueryMap queryMap: BaseRequest): T
    data  class RadioListResponse(val msg: String, val ret: Int, val sub_ret: Int)
}

The compiling is no any problem. But publicRadioListApi.getRadioList<GetRadioListApi.RadioListResponse>(params) always returns  "Network error: java.lang.ClassCastException: com.google.gson.internal.LinkedTreeMap cannot be cast to com.bmwgroup.apinext.music.backend.webapi.radio.GetRadioListApi$RadioGroup"

Is it a valid case?

**Comment by JakeWharton** [comment]
There is no way to access the real value of T which is erased to Object by the compiler. This is why Gson is returning a map (or a list or a scalar), it has no type information.

**Reply by markwang1971** [reply]
thanks for your reply!

**Reply by summit234** [reply]
> There is no way to access the real [value](https://www.servetheinjured.com/) of T which is erased to Object by the compiler. This is why Gson is returning a map (or a list or a scalar), it has no type information.

thanks

**Reply by DAMIT826** [reply]
Thanks
